aten/src/ATen/TensorOptions.h

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <ATen/ScalarType.h>
+
+#include <cstddef>
+
+namespace at {
+struct Type;
+} // namespace at
+
+namespace at {
+
+// should there be multiple kinds of these?
+struct TensorOptions {
+  TensorOptions() = default;
+
+  /* implicit */ TensorOptions(const Type& type);
+
+  TensorOptions& device(size_t device) {
+    device_ = device;
+    return *this;
+  }
+  TensorOptions& backend(Backend backend) {
+    backend_ = backend;
+    return *this;
+  }
+  TensorOptions& dtype(ScalarType dtype) {
+    dtype_ = dtype;
+    return *this;
+  }
+  TensorOptions& requires_grad(bool requires_grad) {
+    requires_grad_ = requires_grad;
+    return *this;
+  }
+
+  size_t device() const noexcept {
+    return device_;
+  }
+  Backend backend() const noexcept {
+    return backend_;
+  }
+  ScalarType dtype() const noexcept {
+    return dtype_;
+  }
+  bool requires_grad() const noexcept {
+    return requires_grad_;
+  }
+
+  /// Constructs an `at::Type` from the members of the `TensorOptions`.
+  Type& type() const;
+
+ private:
+  // shold this be optional?
+  size_t device_ = 0;
```

Comment:
It should. `cuda` without device ordinal means the device currently set in the CUDA context, not device 0.